spec/unit/actions/synchronize_user_filters_spec.rb

```diff
@@ -42,70 +48,90 @@
       allow(project).to receive(:metadata).and_return({})
       allow(project).to receive(:uri)
       allow(data_source).to receive(:realize)
-      allow(organization).to receive(:id).and_return('client123')
+      allow(organization).to receive(:id).and_return('123456789')
       allow(organization).to receive(:project).and_return(project)
     end
     context 'when mode requires client_id' do
       before do
-        allow(domain).to receive(:clients).and_return(organization)
+        allow(domain).to receive(:clients).with(:all, data_product).and_return([organization])
+        allow(domain).to receive(:clients).with('123456789', data_product).and_return(organization)
       end
-      let(:mode) { 'sync_multiple_projects_based_on_custom_id' }
-      let(:params) do
-        params = {
+      let(:params_stub) do
+        {
           GDC_GD_CLIENT: client,
-          input_source: 'foo',
+          input_source: {},
           domain: 'bar',
           filters_config: { labels: [] },
-          sync_mode: mode,
-          gdc_logger: logger
+          gdc_logger: logger,
+          data_product: data_product,
+          multiple_projects_column: 'client_id'
         }
-        GoodData::LCM2.convert_to_smart_hash(params)
       end
 
       it_behaves_like 'a user action reading client_id' do
+        let(:params) { GoodData::LCM2.convert_to_smart_hash(params_stub.merge(sync_mode: 'sync_multiple_projects_based_on_custom_id')) }
         let(:client_id) { '123456789' }
       end
 
-      context 'when mode is sync_domain_client_workspaces' do
-        let(:mode) { 'sync_domain_client_workspaces' }
-        let(:clients) { [] }
+      context 'when using multiple_projects modes' do
+        let(:message_for_project) { :add_data_permissions }
         before do
-          allow(domain).to receive(:clients).with(:all, nil).and_return([organization])
-          allow(domain).to receive(:clients).with(123_456_789).and_return(organization)
+          allow(domain).to receive(:clients).with(:all, data_product).and_return([organization])
+          allow(domain).to receive(:clients).with(123_456_789, data_product).and_return(organization)
           allow(organization).to receive(:project).and_return(project)
           allow(organization).to receive(:client_id).and_return(123_456_789)
           allow(File).to receive(:open).and_return("client_id\n123456789")
           allow(project).to receive(:deleted?).and_return false
         end
 
-        it 'returns results' do
-          result = subject.class.call(params)
-          expect(result).to eq(results: [])
+        context 'sync_multiple_projects_based_on_custom_id mode' do
```

Comment:
warning: Similar code found in synchronize_user_filters_spec.rb:87, synchronize_users_spec.rb:137, synchronize_users_spec.rb:151, synchronize_users_spec.rb:165 (mass = 204) (flay)